README
OpenVPN-ProxyManager

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc OpenVPN::ProxyManager

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=OpenVPN-ProxyManager

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/OpenVPN-ProxyManager

    CPAN Ratings
        http://cpanratings.perl.org/d/OpenVPN-ProxyManager

    Search CPAN
        http://search.cpan.org/dist/OpenVPN-ProxyManager/


LICENSE AND COPYRIGHT

Copyright (C) 2013 David Farrell

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for OpenVPN-ProxyManager

0.01   	3rd February 2013
		First cut:
		ProxyManager  with test_openvpn, create_config, connect methods. 
		Config with print_config
		HMA (ProxyManager superclass) with get_server_list, parse_server_list_string, get_servers, get_hma_config, connect_to_random_server, hma_connect.


lib/Net/OpenVPN/ProxyManager/HMA.pm
package Net::OpenVPN::ProxyManager::HMA;
use LWP::Simple;
use Moose;
use 5.10.0;
extends 'Net::OpenVPN::ProxyManager';

has hma_server_list => (is => 'rw', isa => 'ArrayRef', builder => '_get_server_list');
has hma_config 		=> (is => 'rw', isa => 'Str', builder => '_get_hma_config');

=head1 NAME

OpenVPN::ProxyManager::HMA - connect to HideMyAss.com (HMA) proxy servers using OpenVPN.

=head1 SYNOPSIS

	use Net::OpenVPN::ProxyManager::HMA;
	
	my $pm_hma = Net::OpenVPN::ProxyManager::HMA->new;
	my $servers = $pm_hma->get_servers({ name => 'usa', proto => 'tcp'});
	$pm_hma->connect_to_random_server($servers);
	

=head1 DESCRIPTION

Net::OpenVPN::ProxyManager::HMA is an object oriented module that automatically
downloads the latest list of available HMA proxy servers and the OpenVPN connection
configuration (this is done on construction of the Net::OpenVPN::ProxyManager object).


=head1 Dependencies (non-Perl)

This module has been tested on Ubuntu linux 12.04 with OpenVPN version 2.2.1 built from
Ubuntu package. It should work on other Linux distros, perhaps OSX but probably not
Windows.

To login to the HMA proxy servers, you will need to have an active account with HMA 
(http://hidemyass.com). I am not affiliated with HMA other than as a customer.

You will also need an internet connection!

=cut

sub _get_hma_config {
	my $self = shift;
	my $hma_config_string = get('https://securenetconnection.com/vpnconfig/openvpn-template.ovpn');	
	$hma_config_string ? $hma_config_string : 0;	
}

sub _get_server_list {
	my $self = shift;
	my $hma_server_list_string = get('https://securenetconnection.com/vpnconfig/servers-cli.php');
	$hma_server_list_string ? $self->_parse_server_list_string($hma_server_list_string) : 0;
}

sub _parse_server_list_string {
	my ($self, $hma_server_list_string) = @_;
	my $server_list_arrayhash;
	my @server_list = split qr/\n/, $hma_server_list_string;
	for my $server (@server_list) {
		my @server_data = split qr/\|/, $server;
		push @{$server_list_arrayhash}, {
			'ip' 			=> $server_data[0],
			'name' 			=> $server_data[1],
			'country_code' 	=> $server_data[2],
			'tcp_flag'		=> $server_data[3],
			'udp_flag'		=> $server_data[4],
			'norandom_flag' => $server_data[5],
		};
	}
	return $server_list_arrayhash;
}
=head2 get_servers

This method returns an arrayhash of HMA servers available (the list is downloaded upon 
construction - Net::OpenVPN::ProxyManager::HMA->new). If no arguments are passed to this
method, it will return the entire arrayhash of available servers (approximately 350).

The method accepts two optional string arguments as key value pairs:
-Name, this is a string of the location name. HMA provide a location name string in the 
 format: "Canada, Ontario, Toronto (LOC1 S1)".
-Proto, this is the protocol option and can be either TCP or UDP. Many of the HMA servers 
 accept both protocols.

Example
	my $usa_tcp_servers_arrayhash = $pm_hma->get_servers({name => 'usa', proto => 'tcp'});

=cut

sub get_servers {
	my ($self, $server_params_hashref) = @_;
	my $server_list_arrayhash;
	if (exists $server_params_hashref->{name} and not exists $server_params_hashref->{proto}){
		push @{$server_list_arrayhash},	grep { 
			$_->{name} =~ m/$server_params_hashref->{name}/i} @{$self->hma_server_list};
	}
	elsif (exists $server_params_hashref->{proto} and not exists $server_params_hashref->{name}){
		for ($server_params_hashref->{proto}) {
			when (qr/tcp/i) {
				push @{$server_list_arrayhash},	grep { $_->{tcp_flag}
					} @{$self->hma_server_list};
			}
			when (qr/udp/i) {
				push @{$server_list_arrayhash},	grep { $_->{udp_flag}
					} @{$self->hma_server_list};
			}	
		}
	}
	elsif (exists $server_params_hashref->{proto} and exists $server_params_hashref->{name}){
		for ($server_params_hashref->{proto}) {
			when (qr/tcp/i) {
				push @{$server_list_arrayhash},	grep { $_->{tcp_flag} and 
					$_->{name} =~ m/$server_params_hashref->{name}/i} @{$self->hma_server_list};
			}
			when (qr/udp/i) {
				push @{$server_list_arrayhash},	grep { $_->{udp_flag} and
					$_->{name} =~ m/$server_params_hashref->{name}/i} @{$self->hma_server_list};
			}	
		}
	}
	else {
		push @{$server_list_arrayhash}, $self->hma_server_list;
	}
	return $server_list_arrayhash;
}

=head2 connect_to_random_server

This method will invoke the hma_connect method on a random server when this method is
called with an arrayhash of servers (as returned by the get_servers method).

Example
	$pm_hma->connect_to_random_server($arrayhash_of_servers);

=cut

sub connect_to_random_server {
	my ($self, $server_list_arrayhash) = @_;
	my $server_hashref = $server_list_arrayhash->[int(rand(@{$server_list_arrayhash}-1))];
	print 'Connecting to server ' . $server_hashref->{name} .' '. $server_hashref->{ip} ."\n";
	$self->hma_connect($server_hashref);
}

=head2 hma_connect

The hma_connect method will initialise the OpenVPN program to a server with the HMA
configuration. This method requires a hashref containing the attributes of the
server (this is the same hashref format that is returned by the get_servers method).

Example
	my $pm_hma->hma_connect({
			'ip' 				=> '104.202.33.5',
			'name' 				=> 'Canada, Ontario, Toronto (LOC1 S1)',
			'country_code' 	=> 'ca',
			'tcp_flag'			=> 'TCP',
			'udp_flag'			=> 'UDP',
			'norandom_flag' 	=> undefined
		});

=cut

sub hma_connect {
	my ($self, $server_hashref) = @_;
	return 0 unless defined $server_hashref;
	my $hma_config = $self->hma_config;
	if ($server_hashref->{udp_flag}) {
		$hma_config .= "\n" . 'remote ' . $server_hashref->{ip} . ' 53' .
			"\n" . 'proto udp' . "\n";
	}
	else {
		$hma_config .= "\n" . 'remote ' . $server_hashref->{ip} . ' 443' .
			"\n" . 'proto tcp' . "\n";
	}
	$self->connect($hma_config);
}

no Moose;
1;

lib/Net/OpenVPN/ProxyManager/Config.pm
package Net::OpenVPN::ProxyManager::Config;
use Moose;
use 5.10.0;

has client 			=> (is => 'rw', isa => 'Str', default => '');
has dev 			=> (is => 'rw', isa => 'Str', default => 'tun');
has 'resolv-retry' 	=> (is => 'rw', isa => 'Str', default => 'infinite');
has nobind			=> (is => 'rw', isa => 'Str', default => '');
has 'persist-key'	=> (is => 'rw', isa => 'Str', default => '');
has 'persist-tun'	=> (is => 'rw', isa => 'Str', default => '');
has 'auth-user-pass'=> (is => 'rw', isa => 'Str', default => '');
has 'tun-mtu'		=> (is => 'rw', isa => 'Str', default => '1500');
has 'tun-mtu-extra'	=> (is => 'rw', isa => 'Str', default => '32');
has 'mssfix'		=> (is => 'rw', isa => 'Str', default => '1450');
has 'ns-cert-type'	=> (is => 'rw', isa => 'Str', default => 'server');
has 'verb'			=> (is => 'rw', isa => 'Str', default => '2');
has 'ca'			=> (is => 'rw', isa => 'Str', default => 'skip');
has 'cert'			=> (is => 'rw', isa => 'Str', default => 'skip');
has 'key'			=> (is => 'rw', isa => 'Str', default => 'skip');
has 'remote'		=> (is => 'rw', isa => 'Str', default => 'skip');
has 'proto'			=> (is => 'rw', isa => 'Str', default => 'skip');

sub print_config {
	my ($self, $fh, $config_string) = @_;
	if (ref $fh ne 'GLOB') {
		print 'Error: print_config method call missing a filehandle parameter'."\n";
		return 0;
	}
	if (not defined $config_string) { # if no config_string arg was supplied
		for my $attribute ($self->meta->get_all_attributes) {
			my $attribute_value = $attribute->get_value($self);
			for ($attribute_value) {	
				when (qr/^skip$/) { next }
				when (qr/\n/) { 
					$config_string.= 
						"<".$attribute->name.">\n".
						$attribute->get_value($self)."\n".
						"</".$attribute->name.">\n"}
				default { $config_string.= $attribute->name.' '.$attribute->get_value($self)."\n" }
			}
		}
	}
	print $fh $config_string;
}

no Moose;
1;

lib/Net/OpenVPN/ProxyManager.pm
package Net::OpenVPN::ProxyManager;
use Moose;
use 5.10.0;
use Net::OpenVPN::ProxyManager::Config;
use Capture::Tiny 'capture';

our $VERSION = '0.01';

has config_path 	=> ( is => 'rw', isa => 'Str', default => '/tmp/openvpn-config.conf' );
has config			=> ( is => 'rw', isa => 'Object', builder => 'create_config');

sub test_openvpn {
	my ($self) = @_;
	my  ($stdout, $stderr, @result) = capture { system 'which', 'openvpn'; };
	$result[0] ? die qq!openvpn not found (are you using Linux?)! : 1;
	chomp $stdout;
	return $stdout;
}

sub create_config {
	my ($self, $params) = @_;
	ref $params eq 'HASH' ? Net::OpenVPN::ProxyManager::Config->new($params) : Net::OpenVPN::ProxyManager::Config->new;
}

sub connect {
	my ($self, $config_string) = @_;
	
	#check openvpn is installed
	my $openvpn_path = $self->test_openvpn;
	
	# open FH and print config file to /tmp
	open (my $fh, '>', $self->config_path() ); 
	$self->config()->print_config($fh, $config_string);
	$fh->close;
	system 'chmod', '644', $self->config_path;
	
	# run openvpn
	system 'sudo', $openvpn_path, '--config', $self->config_path;
}

no Moose;
1;

__END__

=pod

=head1 NAME

Net::OpenVPN::ProxyManager - connect to proxy servers using OpenVPN.

=head1 SYNOPSIS

	use Net::OpenVPN::ProxyManager;
	
	my $pm = Net::OpenVPN::ProxyManager->new;
	
	# Create a config object to enter proxy details
	my $config_object = $pm->create_config({remote => '100.120.3.34 53', proto => 'udp'});
	
	#Launch OpenVPN and connect to the proxy
	$pm->connect($config_object);

=head1 DESCRIPTION

Net::OpenVPN::ProxyManager is an object oriented module that provides methods to
simplify the management of proxy connections that support OpenVPN. This is a base 
generic class, see Net::OpenVPN::ProxyManager::HMA for additional methods to 
interact with hidemyass.com proxy servers. 

=head1 DEPENDENCIES

Non-Perl dependencies

OS - this module has been tested on Ubuntu linux 12.04. It should work on other Linux distros, 
perhaps OSX but probably not Windows.

OpenVPN - this module has been tested on OpenVPN version 2.2.1. I do not know if it will work 
on earlier versions. See Net::OpenVPN::Manage for a Perl module that does not require OpenVPN
to be installed.

You will also need an internet connection!

=cut

=head1 TO DO / UNSATISFACTORY STUFF

By default OpenVPN runs as root and this module uses the 'sudo' command
when starting OpenVPN. Ergh.

The module only enables connection to one proxy server per invocation.
OpenVPN only supports one connection per client, however I would like to change this to a 
forked model where connections can be initialised and dropped without closing the whole process 
via a 'disconnect' method.

=head1 AUTHOR

David Farrell, C<< <davidnmfarrell at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-openvpn-proxymanager at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=OpenVPN-ProxyManager>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc OpenVPN::ProxyManager


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=OpenVPN-ProxyManager>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/OpenVPN-ProxyManager>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/OpenVPN-ProxyManager>

=item * Search CPAN

L<http://search.cpan.org/dist/OpenVPN-ProxyManager/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2013 David Farrell.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

